File: output/0113.路径总和II/0113-路径总和II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# class Solution:
#     def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
class Solution:
    # res = []
    def pathSum(self, root: TreeNode, targetSum: int):
        # self.res = []
        res = []
        path = []
        # @functools.lru_cache(None)
        def dfs(p, val):
            if p.val == val and not p.left and not p.right:
                res.append(path[:] + [p.val])
            # elif val < p.val :
            #     return
            else:
                path.append(p.val)
                if p.left:
                    dfs(p.left, val - p.val)
                if p.right:
                    dfs(p.right, val - p.val)
                path.pop()
        if root:
            dfs(p=root, val=targetSum)
        return res

File: 答题/标签/树/0113.路径总和II/0113-路径总和II.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, targetSum int) [][]int {
    res := [][]int{}
    var dfs func(cur *TreeNode, sum int, l []int)
    dfs = func(cur *TreeNode, sum int, l []int){
        if cur == nil{
            return
        }
        if cur.Left == nil && cur.Right == nil{
            l = append(l, cur.Val)
            if sum + cur.Val == targetSum{
                res = append(res, append([]int{}, l...))
            }
            l = l[:len(l) - 1]
            return
        }
        l = append(l, cur.Val)
        dfs(cur.Left, sum + cur.Val, l)
        dfs(cur.Right, sum + cur.Val, l)
        l = l[:len(l) - 1]
    }
    dfs(root, 0, []int{})
    return res
}